internal/monitors/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,276 @@
+
+		dps := client.GetNodeStatsDatapoints(nodeStatsOutput, nodeMetricsDims, utils.StringSliceToMap(conf.ThreadPools), utils.StringSliceToMap(conf.EnableEnhancedNodeStatsForIndexGroups), map[string]bool{
+			client.HTTPStatsGroup:       conf.EnableEnhancedHTTPStats,
+			client.JVMStatsGroup:        conf.EnableEnhancedJVMStats,
+			client.ProcessStatsGroup:    conf.EnableEnhancedProcessStats,
+			client.ThreadpoolStatsGroup: conf.EnableEnhancedThreadPoolStats,
+			client.TransportStatsGroup:  conf.EnableEnhancedTransportStats,
+		})
+
+		// If current node is not the master, collect only node level stats
+		if !nodeIsCurrentMaster {
+			m.sendDatapoints(dps)
+			return
+		}
+
+		// Collect cluster level stats
+		clusterStatsOutput, err := esClient.GetClusterStats()
+
+		if err != nil {
+			logger.WithError(err).Errorf("Failed to GET cluster stats")
+			return
+		}
+
+		dps = append(dps, client.GetClusterStatsDatapoints(clusterStatsOutput, defaultDimensions, conf.EnableEnhancedClusterHealthStats)...)
+
+		if !*conf.EnableIndexStats {
+			m.sendDatapoints(dps)
+			return
+		}
+
+		// Collect Index stats
+		if !conf.IndexStatsMasterOnly || nodeIsCurrentMaster {
+			indexStatsOutput, err := esClient.GetIndexStats()
+
+			if err != nil {
+				logger.WithError(err).Errorf("Failed to GET index stats")
+				return
+			}
+
+			if conf.IndexSummaryOnly {
+				dps = append(dps, client.GetIndexStatsSummaryDatapoints(indexStatsOutput.AllIndexStats, defaultDimensions, utils.StringSliceToMap(conf.EnableEnhancedIndexStatsForIndexGroups), conf.EnableIndexStatsPrimaries)...)
+				m.sendDatapoints(dps)
+				return
+			}
+
+			dps = append(dps, client.GetIndexStatsDatapoints(indexStatsOutput.Indices, utils.StringSliceToMap(conf.Indexes), defaultDimensions, utils.StringSliceToMap(conf.EnableEnhancedIndexStatsForIndexGroups), conf.EnableIndexStatsPrimaries)...)
+		}
+
+		m.sendDatapoints(dps)
+	}, time.Duration(conf.IntervalSeconds)*time.Second)
+	return nil
+}
+
+// Prepares dimensions that are common to all datapoints from the monitor
+func prepareDefaultDimensions(userProvidedClusterName string, queriedClusterName *string) (map[string]string, error) {
+	dims := map[string]string{}
+	clusterName := userProvidedClusterName
+
+	if clusterName == "" {
+		if queriedClusterName == nil {
+			return nil, errors.New("Failed to GET cluster name from Elasticsearch API")
+		}
+		clusterName = *queriedClusterName
+	}
+
+	// "plugin_instance" dimension is added to maintain backwards compatibility with built-in content
+	dims["plugin_instance"] = clusterName
+	dims["cluster"] = clusterName
+	dims["plugin"] = monitorType
+
+	return dims, nil
+}
+
+func isCurrentMaster(nodeID string, masterInfoOutput *client.MasterInfoOutput) (bool, error) {
+	var isCurrentMaster bool
+
+	masterNode := masterInfoOutput.MasterNode
+
+	if masterNode == nil {
+		logger.Warn("Unable to identify Elasticsearch cluster master node. Assuming current node is not the current master.")
```

Comment:
Use a throttled logger to log this since it is going to happen every collection cycle if it happens at all: https://github.com/signalfx/signalfx-agent/blob/master/internal/utils/log.go#L142

Reply:
Updated